compiler/rustc_resolve/src/lib.rs

```diff
@@ -3392,6 +3403,38 @@ impl<'a> Resolver<'a> {
         }
         None
     }
+
+    fn resolve_main_by_path(&mut self, krate: &ast::Crate) {
+        let any_exe = self.session.crate_types().iter().any(|ty| *ty == CrateType::Executable);
+        if !any_exe {
+            // No need to find a main function.
+            return;
+        }
+
+        // If the user wants no main function at all, then stop here.
+        if self.session.contains_name(&krate.attrs, sym::no_main) {
+            return;
+        }
+
+        let seg_crate =
+            PathSegment { ident: Ident::with_dummy_span(kw::Crate), id: DUMMY_NODE_ID, args: None };
+        let seg_main =
+            PathSegment { ident: Ident::with_dummy_span(sym::main), id: DUMMY_NODE_ID, args: None };
+        let path = Path { span: DUMMY_SP, segments: vec![seg_crate, seg_main], tokens: None };
+
+        let module = self.get_module(LocalDefId { local_def_index: CRATE_DEF_INDEX }.to_def_id());
+        let parent_scope = &ParentScope::module(module, self);
+        let res = match self.resolve_ast_path(&path, Namespace::ValueNS, parent_scope, true).ok() {
```

Comment:
I suggest using `resolve_ident_in_module` instead.
Then you'll get `NameBinding` as a result, which will tell you whether the `main` is imported or not, which is needed for feature gating.
You'll also be able to mark the result as used (with `record_use`) only if it is a function.

Reply:
Yes, this is much better! Thanks!